Check if the buffers fit GPU memory after device map auto inferred

notsyncing opened the pull request:
# What does this PR do?

Hello, I'm trying using `accelerate` to offload a large model (https://huggingface.co/TheBloke/WizardCoder-33B-V1.1-GPTQ) to CPU, with following code (requires https://github.com/huggingface/accelerate/pull/2383 if using Intel GPU, and https://github.com/huggingface/transformers/pull/28755):

```python
import datetime

import torch
import intel_extension_for_pytorch

import accelerate
from transformers import AutoModelForCausalLM, AutoTokenizer, pipeline

model_id = "/mnt/external2/LLMs/WizardCoder-33B-V1.1-GPTQ"
tokenizer = AutoTokenizer.from_pretrained(model_id)

# Without offload_buffers (OOM)
pipe = pipeline("text-generation", model=model_id, model_kwargs={"torch_dtype": torch.bfloat16, "device_map": "auto", "max_memory": {0: "16GB", "cpu": "128GB"}})

# With offload_buffers (won't OOM)
#pipe = pipeline("text-generation", model=model_id, model_kwargs={"torch_dtype": torch.bfloat16, "device_map": "auto", "max_memory": {0: "16GB", "cpu": "128GB"}, "offload_buffers": True})
print(str(pipe.model.hf_device_map))

print(str(datetime.datetime.now()) + " Generating...")
results = pipe("public void helloWorld() {")

print(str(datetime.datetime.now()) + " Output:")
print(results)
``` 

I have one Intel Arc A770 16GB GPU in my machine, but the code above always OOM on the GPU.

After some digging, I found that this model `TheBloke/WizardCoder-33B-V1.1-GPTQ` contains a huge buffer (model is 17GB, but buffers are more than 16GB), so it will not fit into my GPU at all without `offload_buffers`.

So I created this PR, to check if the buffer can fit on the GPU, and raise an exception for such case.



## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/accelerate/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/accelerate/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/accelerate/tree/main/docs#writing-documentation---specification).
- [x] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.

- Big modeling: @SunMarc
- Fully-Sharded Data Parallism: @pacman100
- DeepSpeed: @pacman100
- Command Line Interface: @muellerzr
- Documentation: @muellerzr
- Core parts of the library: @muellerzr @BenjaminBossan
- Maintained examples: @muellerzr or @pacman100

 -->

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/accelerate/pr_2412). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

SunMarc commented on src/accelerate/utils/modeling.py:
```diff
@@ -1267,8 +1308,33 @@ def infer_auto_device_map(
+                continue
+
+            gpu_memory_used = device_memory_used.get(gpu_device, 0)
+            if gpu_max_memory < non_gpu_buffer_size + gpu_memory_used:
+                raise ValueError(
+                    f"Device {gpu_device} has {gpu_max_memory} bytes of memory, and has planned to use {gpu_memory_used} bytes, "
+                    f"so the remaining memory is not enough to store additional buffers ({non_gpu_buffer_size} bytes)! "
+                    f"Please consider using offload_buffers=True."
+                )
```
This will trigger an error even when we have, for example, the first gpu being free but the second gpu is full. The buffers can potentially go on the first gpu. When we don't offload the buffers, they are on the execution device (can be any device). Maybe we can just put a warning to let the user know that the buffer that won't be offloaded and it won't probably fit on one of the gpus.

notsyncing replied:
Yes, I did not consider the multi-gpu scenario because I have only one card. I'll change this to a warning instead of an `Error`. I will also try to determine the execution device more precisely to see if we can make this warning be more useful.

notsyncing replied:
btw, could you please give some hint on the code which implemented the "the first gpu being free but the second gpu is full. The buffers can potentially go on the first gpu"? The only thing I found is the `AlignDevicesHook` class in `src/accelerate/hooks.py`, which seems moving all buffers to all execution devices.

SunMarc replied:
Basically, what I mean is if the user let a `device_map` that only most of the modules on the second gpus and leave of the space available on the first gpu, the buffer will most likely fit in the first gpu (if we suppose the first gpu is the execution device). For example, this is the case with `device_map="balanced_low_0"` strategy.

SunMarc reviewed (commented):
Thanks for digging into the buffer issue @notsyncing ! Could you add some tests to see that the buffer calculation is right and that we raise the error when the buffers don't fit the remaining space on the gpu. Can you have a second look @muellerzr ?

notsyncing commented on src/accelerate/utils/modeling.py:
```diff
@@ -698,7 +699,16 @@ def compute_module_sizes(
+    if not buffers_only:
+        module_list = named_module_tensors(model, recurse=True) 
+    else:
+        if hasattr(model, "named_buffers"): 
```
But there is a test that won't not pass without this. I'm not sure if I should touch the code of that test.

SunMarc replied:
Could you link the test ?

notsyncing replied:
```
FAILED tests/test_modeling_utils.py::ModelingUtilsTester::test_infer_auto_device_map_with_tied_weights - AttributeError: 'Parameter' object has no attribute 'named_buffers'
``` 

detail log:

```
_________________________ ModelingUtilsTester.test_infer_auto_device_map_with_tied_weights _________________________

self = <test_modeling_utils.ModelingUtilsTester testMethod=test_infer_auto_device_map_with_tied_weights>

    def test_infer_auto_device_map_with_tied_weights(self):
        model = nn.Sequential(
            OrderedDict([("layer1", ModelForTest()), ("layer2", ModelForTest()), ("layer3", ModelForTest())])
        )
        model.layer3.linear2.weight = model.layer1.linear2.weight
        device_map = infer_auto_device_map(model, max_memory={0: 400, 1: 500})
        expected = {"layer1": 0, "layer3.linear2": 0, "layer2": 1, "layer3.linear1": 1, "layer3.batchnorm": 1}
        assert device_map == expected
    
        # With three weights tied together
        model.layer2.linear2.weight = model.layer1.linear2.weight
        device_map = infer_auto_device_map(model, max_memory={0: 400, 1: 500})
        expected = {
            "layer1": 0,
            "layer2.linear2": 0,
            "layer3.linear2": 0,
            "layer2.linear1": 1,
            "layer2.batchnorm": 1,
            "layer3.linear1": 1,
            "layer3.batchnorm": 1,
        }
        assert device_map == expected
    
        # With two groups of weights tied together
        model.layer2.linear1.weight = model.layer1.linear1.weight
        device_map = infer_auto_device_map(model, max_memory={0: 400, 1: 500})
        expected = {
            "layer1": 0,
            "layer2.linear1": 0,
            "layer2.linear2": 0,
            "layer3.linear2": 0,
            "layer2.batchnorm": 1,
            "layer3.linear1": 1,
            "layer3.batchnorm": 1,
        }
        assert device_map == expected
    
        # With weights ties in the same module
        model = nn.Sequential(
            OrderedDict(
                [
                    ("linear1", nn.Linear(4, 4)),
                    ("linear2", nn.Linear(6, 6)),
                    ("linear3", nn.Linear(4, 4)),
                    ("linear4", nn.Linear(6, 6)),
                ]
            )
        )
        model.linear3.weight = model.linear1.weight
        model.linear3.bias = model.linear1.bias
        device_map = infer_auto_device_map(model, max_memory={0: 250, 1: 400})
        expected = {"linear1": 0, "linear2": 1, "linear3": 0, "linear4": 1}
        assert device_map == expected
    
        # With tied weights sharing a same prefix name (`compute.weight` vs `compute.weight_submodule.parameter`)
        class SubModule(torch.nn.Module):
            def __init__(self, ref_to_parameter):
                super().__init__()
                self.parameter = ref_to_parameter
    
            def forward(self, x):
                return self.x + torch.max(self.parameter)
    
        class LinearModuleAndSubModule(torch.nn.Linear):
            def __init__(self, in_features, out_features):
                super().__init__(in_features, out_features)
                self.weight_submodule = SubModule(self.weight)
    
            def forward(self, x):
                return torch.nn.functional.linear(self.weight_submodule(x), self.weight)
    
        class Model(torch.nn.Module):
            def __init__(self):
                super().__init__()
                self.compute = LinearModuleAndSubModule(3, 8)
    
            def forward(self, x):
                return self.compute(x)
    
        model = Model()
    
        device_memory = {0: 4, "cpu": 96000}  # Low memory device, just to force splitting and trigger the error
>       infer_auto_device_map(model, device_memory)

tests/test_modeling_utils.py:581: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
src/accelerate/utils/modeling.py:1251: in infer_auto_device_map
    current_buffer_size = compute_module_total_buffer_size(
src/accelerate/utils/modeling.py:740: in compute_module_total_buffer_size
    module_sizes = compute_module_sizes(model, dtype, special_dtypes, True)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

model = Parameter containing:
tensor([[-0.1732, -0.2755,  0.5513],
        [ 0.3202, -0.3520, -0.2377],
        [ 0.1762,  0.4...0662, -0.2712, -0.3916],
        [-0.3685,  0.5621,  0.1996],
        [ 0.2687,  0.0723, -0.5561]], requires_grad=True)
dtype = None, special_dtypes = None, buffers_only = True

    def compute_module_sizes(
        model: nn.Module,
        dtype: Optional[Union[str, torch.device]] = None,
        special_dtypes: Optional[Dict[str, Union[str, torch.device]]] = None,
        buffers_only: bool = False,
    ):
        """
        Compute the size of each submodule of a given model.
        """
        if dtype is not None:
            dtype = _get_proper_dtype(dtype)
            dtype_size = dtype_byte_size(dtype)
        if special_dtypes is not None:
            special_dtypes = {key: _get_proper_dtype(dtyp) for key, dtyp in special_dtypes.items()}
            special_dtypes_size = {key: dtype_byte_size(dtyp) for key, dtyp in special_dtypes.items()}
        module_sizes = defaultdict(int)
    
        module_list = []
    
        if not buffers_only:
            module_list = named_module_tensors(model, recurse=True)
        else:
            #if hasattr(model, "named_buffers"):
>           module_list = model.named_buffers(recurse=True)
E           AttributeError: 'Parameter' object has no attribute 'named_buffers'

src/accelerate/utils/modeling.py:712: AttributeError
```

SunMarc replied:
I found the issue. This happens because you don't make sure that the `module` that you are passing in `infer_auto_device_map` is indeed a `module`. See comment below ! If the changes, it should be safe to remove the condition.

SunMarc commented:
Hi @notsyncing, thanks for your work ! I've replied to your questions. This is not pressing but are you planning to finish this PR ? Otherwise, I can finish it and add you as co-author =)

notsyncing commented:
> Hi @notsyncing, thanks for your work ! I've replied to your questions. This is not pressing but are you planning to finish this PR ? Otherwise, I can finish it and add you as co-author =)

Yes, I'm planning to finish it. But for the multi-gpu scenario, I cannot test it because I have only one gpu. Would you mind helping testing this? Thanks!

muellerzr commented:
We can run the relevant tests for you when you are ready/the normal CI here is green and report back any failures, yes :)

notsyncing commented on src/accelerate/utils/modeling.py:
```diff
@@ -1211,6 +1241,15 @@ def infer_auto_device_map(
+                        module, dtype=dtype, special_dtypes=special_dtypes
+                    )
+                    device_buffer_sizes[device] = device_buffer_sizes[device] + current_buffer_size
+
```
Fixed.

notsyncing commented on src/accelerate/utils/modeling.py:
```diff
@@ -1211,6 +1241,15 @@ def infer_auto_device_map(
 
+                if not offload_buffers:
+                    if device not in device_buffer_sizes.keys():
+                        device_buffer_sizes[device] = 0
```
Fixed

notsyncing commented on src/accelerate/utils/modeling.py:
```diff
@@ -716,6 +726,18 @@ def compute_module_sizes(
+    """
+    Compute the total size of buffers in each submodule of a given model.
+    """
+    module_sizes = compute_module_sizes(model, dtype, special_dtypes, True)
```
Fixed

notsyncing commented on src/accelerate/utils/modeling.py:
```diff
@@ -1267,8 +1308,33 @@ def infer_auto_device_map(
 
+            if not offload_buffers:
+                if device not in device_buffer_sizes.keys():
+                    device_buffer_sizes[device] = 0
```
Fixed

notsyncing commented on src/accelerate/utils/modeling.py:
```diff
@@ -1267,8 +1308,33 @@ def infer_auto_device_map(
+                    module, dtype=dtype, special_dtypes=special_dtypes
+                )
+                device_buffer_sizes[device] = device_buffer_sizes[device] + current_buffer_size
+
```
Fixed

notsyncing commented:
> Basically, what I mean is if the user let a `device_map` that only most of the modules on the second gpus and leave of the space available on the first gpu, the buffer will most likely fit in the first gpu (if we suppose the first gpu is the execution device). For example, this is the case with `device_map="balanced_low_0"` strategy.

Now I added a check to see if the buffer can fit any GPU, and only print the warning when every GPU does not fit. Will this work in your scenario?

notsyncing commented:
> Could you add some tests to see that the buffer calculation is right and that we raise the error when the buffers don't fit the remaining space on the gpu. 

As for the tests, it seems difficult to write a test for this if it only prints a warning. I propose that we could use a environment variable (like `ACCELERATE_BUFFER_SIZE_CHECK = "raise"`) to make it raises an exception instead:

```python
if os.getenv("ACCELERATE_BUFFER_SIZE_CHECK") == "raise":
    raise ValueError("xxx")
else:
    logger.warn("yyy")
```

Would this be better?

muellerzr commented:
You can use `warnings.warning()` to do so, and unittest has a method to check for a warning: `with self.assertWarns(Warning):`

muellerzr commented:
Also there's no need to force push, we squash the commit history when merging

SunMarc commented:
> Now I added a check to see if the buffer can fit any GPU, and only print the warning when every GPU does not fit. Will this work in your scenario?

Thanks ! Let's do that first and see the feedback. We can always change this behavior if needed. Most probably, the user will receive this warning if all his gpus are full and some of the layers are offloaded to cpu/disk. Hence, the buffer won't fit any of the gpus.

SunMarc commented on src/accelerate/utils/modeling.py:
```diff
@@ -1214,6 +1244,12 @@ def infer_auto_device_map(
                         modules_to_treat.pop(tied_module_index)
                     device_map[tied_module_name] = devices[current_device]
 
+                if not offload_buffers:
```
We need to add an extra condition to make sure that we are indeed passing a `nn.Module` in `compute_module_total_buffer_size` since `module` can also be a `nn.Parameter`. This is why the test was failing.

notsyncing replied:
fixed

SunMarc commented on src/accelerate/utils/modeling.py:
```diff
@@ -1270,8 +1308,34 @@ def infer_auto_device_map(
             current_memory_used += module_size
             device_map[name] = devices[current_device]
 
+            if not offload_buffers:
```
same here !

notsyncing replied:
fixed

SunMarc reviewed (commented):
Thanks for the iteration. I left a few comments. Could you also add tests that would trigger the warning for a single and multi gpu setup ? We can test the multi gpu test if needed !

notsyncing commented:
@SunMarc I have added two test cases in `test_modeling_utils.py`, and fixed a problem that my previous code did not consider the reserved memory for the largest layer. Looks like we don't need an actual multi-gpu setup to test this.

SunMarc commented on tests/test_modeling_utils.py:
```diff
@@ -604,6 +618,59 @@ def test_infer_auto_device_map_on_t0pp(self):
+        # Now we have two devices, linear1 will fit on device 0, batchnorm will fit on device 1, and the second device
+        # can hold all remaining buffers
+        # Should NOT print a warning in such case
+        with warnings.catch_warnings():
+            warnings.simplefilter("error")
+            device_map = infer_auto_device_map(model, max_memory={0: 400, 1: 400, "cpu": "1GB"})
+        assert device_map == {"linear1": 0, "batchnorm": 1, "linear2": "cpu", "linear3": "cpu"}
```
We need to use simplefilter("always") and we can test if we catched any warning 
```suggestion
        with warnings.catch_warnings(record=True) as w:
            warnings.simplefilter("always")
            device_map = infer_auto_device_map(model, max_memory={0: 400, 1: 400, "cpu": "1GB"})
        assert len(w) == 0
        assert device_map == {"linear1": 0, "batchnorm": 1, "linear2": "cpu", "linear3": "cpu"}
```

notsyncing replied:
fixed

SunMarc commented on tests/test_modeling_utils.py:
```diff
@@ -604,6 +618,59 @@ def test_infer_auto_device_map_on_t0pp(self):
+
+        # Now we have two devices, neither can hold all the buffers, but we are using the offload_buffers=True
+        # Should NOT print a warning in such case
+        with warnings.catch_warnings():
+            warnings.simplefilter("error")
+            device_map = infer_auto_device_map(model, max_memory={0: 400, 1: 200, "cpu": "1GB"}, offload_buffers=True)
```
```suggestion
        with warnings.catch_warnings(record=True) as w:
            warnings.simplefilter("always")
            device_map = infer_auto_device_map(model, max_memory={0: 400, 1: 200, "cpu": "1GB"}, offload_buffers=True)
        assert len(w) == 0
```

notsyncing replied:
fixed

SunMarc commented on tests/test_modeling_utils.py:
```diff
@@ -604,6 +618,59 @@ def test_infer_auto_device_map_on_t0pp(self):
+        # Only linear1 (144) fits on device 0, and remaining buffers (batchnorm's 160 + linear2's 120 = 280) won't fit
+        # device 0, but with offload_buffers they won't be loaded to device 0 all at once, so it's ok now
+        # Should NOT print a warning in such case
+        with warnings.catch_warnings():
+            warnings.simplefilter("error")
+            device_map = infer_auto_device_map(model, max_memory={0: 400, "cpu": "1GB"}, offload_buffers=True)
```
```suggestion
        with warnings.catch_warnings(record=True) as w:
            warnings.simplefilter("always")
            device_map = infer_auto_device_map(model, max_memory={0: 400, "cpu": "1GB"}, offload_buffers=True)
        assert len(w) == 0
```

notsyncing replied:
fixed

SunMarc replied:
Hi @notsyncing, is there a reason why you didn't to add the `check assert len(w) == 0` with `with warnings.catch_warnings(record=True) as w `: ?  We need that since we don't really pay attention to the warning that show up in the logs

notsyncing replied:
Oh, sorry, I missed this line :joy:, will add it later

SunMarc reviewed (commented):
Thanks for working on this PR and adding the tests ! I ran them and they passed. I left a couple of minor comments/suggestions. We can merge this PR after fixing these =)

notsyncing commented on src/accelerate/utils/modeling.py:
```diff
@@ -1098,6 +1124,9 @@ def infer_auto_device_map(
     current_device = 0
     current_memory_used = 0
+    device_memory_used = {}
+    device_memory_reserved = {}
```
Removed `device_memory_reserved` and using a `current_memory_reserved` in each loop instead.

notsyncing commented on src/accelerate/utils/modeling.py:
```diff
@@ -1272,10 +1312,39 @@ def infer_auto_device_map(
         device_map = clean_device_map(device_map)
+
+    if not offload_buffers:
+        non_gpu_buffer_size = device_buffer_sizes.get("cpu", 0) + device_buffer_sizes.get("disk", 0)
```
fixed.

notsyncing commented on src/accelerate/utils/modeling.py:
```diff
@@ -1150,6 +1179,7 @@ def infer_auto_device_map(
         # Reduce max size available by the largest layer.
         if devices[current_device] in main_devices:
             current_max_size = current_max_size - max_layer_size
+            device_memory_reserved[current_device] = max_layer_size
```
fixed

notsyncing commented:
> Thanks for working on this PR and adding the tests ! I ran them and they passed. I left a couple of minor comments/suggestions. We can merge this PR after fixing these =)

@SunMarc all fixed.

SunMarc reviewed (approved):
Thanks for iterating ! LGTM. I re-opened a comment I made. Could you have a quick look ? Also, can you have a second look @muellerzr ?

muellerzr reviewed (approved):
Thanks! Overall this looks sound to me :)
